Fix new Crossfeed DNS entries

epicfaace opened the pull request:
Fix new Crossfeed DNS entries by doing the following:
- Pointing the existing A and AAAA records for Cloudfront Distribution on the CD AWS account to the new domains, `staging-cd.crossfeed.cyber.dhs.gov` and `api.staging-cd.crossfeed.cyber.dhs.gov`
- I created new ACM certificates for `staging-cd.crossfeed.cyber.dhs.gov` and `api.staging-cd.crossfeed.cyber.dhs.gov` and added the corresponding DNS validation CNAME records in the terraform code so that AWS can issue these certificates.
- Re-added ACM certificate DNS validation CNAME records for the SSL certs for domains `staging.crossfeed.cyber.dhs.gov` and `api.staging.crossfeed.cyber.dhs.gov` (which were mistakenly removed in https://github.com/cisagov/cool-dns-cyber.dhs.gov/pull/76)
- Renaming `api-cd.staging.crossfeed.cyber.dhs.gov` to `api.staging-cd.crossfeed.cyber.dhs.gov` for naming uniformity (`api.[STAGE].crossfeed.cyber.dhs.gov`)
- Added A / AAAA records for `staging.crossfeed.cyber.dhs.gov` and `api.staging.crossfeed.cyber.dhs.gov` to point to the LZ.

For broader context, this change is to enable `staging.crossfeed.cyber.dhs.gov` and `api.staging.crossfeed.cyber.dhs.gov` to point to the load balancers in the LZ account in GovCloud, while `staging-cd.crossfeed.cyber.dhs.gov` and `api.staging-cd.crossfeed.cyber.dhs.gov` will point to the existing deployed Crossfeed infrastructure in the CD commercial cloud environment. This way, the migration to LZ will not block development work from continuing.

epicfaace commented:
@mcdonnnj @jsf9k If you could run `terraform-docs` on this PR to update the README, that would be great! I tried running `terraform-docs` but I seem to be using a newer version that's overwriting a lot of the existing formatting -- don't know which version generates the current README file.

mcdonnnj commented:
> @mcdonnnj @jsf9k If you could run `terraform-docs` on this PR to update the README, that would be great! I tried running `terraform-docs` but I seem to be using a newer version that's overwriting a lot of the existing formatting -- don't know which version generates the current README file.

@epicfaace Did you follow the command usage provided in https://github.com/cisagov/cool-dns-cyber.dhs.gov/pull/76#discussion_r1269955409?

epicfaace commented:
> > @mcdonnnj @jsf9k If you could run `terraform-docs` on this PR to update the README, that would be great! I tried running `terraform-docs` but I seem to be using a newer version that's overwriting a lot of the existing formatting -- don't know which version generates the current README file.
> 
> @epicfaace Did you follow the command usage provided in [#76 (comment)](https://github.com/cisagov/cool-dns-cyber.dhs.gov/pull/76#discussion_r1269955409)?

Oh didn’t see that earlier, thanks. I’m out of office the rest of the week and not at a computer until Monday, if it isn’t too much trouble could you run the command?

jsf9k commented on route53_crossfeed_app.tf:
```diff
@@ -243,6 +243,16 @@ resource "aws_route53_record" "crossfeed_staging_CNAME2" {
   zone_id = aws_route53_zone.cyber_dhs_gov.zone_id
 }
```
I can't apply this `CNAME` record because there exist other non-`CNAME` records for `staging.crossfeed.cyber.dhs.gov`.

lfilip-ivaal replied:
@jsf9k  What can we do for this issue? I can replicate these changes with any alterations needed in a new PR from one of my own forks if needed.  Not sure what exactly needs to change.

jsf9k replied:
I'm not sure what needs to be done.  We can just wait for @epicfaace to return on Monday.

epicfaace replied:
Can you explain more why this is an issue? Why would non-CNAME records on the same domain prevent adding a CNAME record to it?

jsf9k replied:
[It's a fundamental constraint due to the way `CNAME` records are handled in DNS](https://serverfault.com/questions/853481/cant-add-dns-cname-record#:~:text=A%20CNAME%20is%20just%20an,touch%20your%20DNS%20for%20this!).

stewartl97 replied:
@jsf9k the records you are referring to then are the MX and TXT records that are also listed?  Is it just a matter of moving all of the records to the new domain?

jsf9k replied:
@stewartl97 - It depends how you choose to fix it.  You can get rid of the `CNAME` record and explicitly list all records, or you can keep the `CNAME` and delete the other records.  I don't know what is best for your use case.

jsf9k commented on route53_crossfeed_app.tf:
```diff
@@ -307,11 +317,12 @@ resource "aws_route53_record" "crossfeed_staging_api_CNAME1" {
   zone_id = aws_route53_zone.cyber_dhs_gov.zone_id
 }
 
+
```
Remove needless blank line:
```suggestion
```

jsf9k reviewed (changes requested):
Please note that I squashed your fixup commits into the original.

epicfaace commented:
@jsf9k I've updated this PR by using alias records instead of CNAME records to point to the NLB. Let me know if this works. Thanks!

jsf9k commented:
> @jsf9k I've updated this PR by using alias records instead of CNAME records to point to the NLB. Let me know if this works. Thanks!

@epicfaace - I tried to apply the latest changes, but I get these errors:
```
╷
│ Error: Error describing created certificate: Expected certificate to be issued but was in state EXPIRED
│
│   with aws_acm_certificate_validation.rules,
│   on acm_rules_vm.tf line 49, in resource "aws_acm_certificate_validation" "rules":
│   49: resource "aws_acm_certificate_validation" "rules" {
│
╵
╷
│ Error: [ERR]: Error building changeset: InvalidChangeBatch: [Tried to create an alias that targets crossfeed-stage-1792947306.us-gov-west-1.elb.amazonaws.com., type A in zone ZMG1MZ2THAWF1, but the alias target name does not lie within the target zone]
│       status code: 400, request id: 2943d76b-fbbe-4c60-a49f-f2af3d9013f0
│
│   with aws_route53_record.crossfeed_staging_A,
│   on route53_crossfeed_app.tf line 229, in resource "aws_route53_record" "crossfeed_staging_A":
│  229: resource "aws_route53_record" "crossfeed_staging_A" {
│
╵
╷
│ Error: [ERR]: Error building changeset: InvalidChangeBatch: [Tried to create an alias that targets crossfeed-stage-1792947306.us-gov-west-1.elb.amazonaws.com., type AAAA in zone ZMG1MZ2THAWF1, but the alias target name does not lie within the target zone]
│       status code: 400, request id: 8384b67c-eb20-4e37-aa96-08a218db5d5d
│
│   with aws_route53_record.crossfeed_staging_AAAA,
│   on route53_crossfeed_app.tf line 242, in resource "aws_route53_record" "crossfeed_staging_AAAA":
│  242: resource "aws_route53_record" "crossfeed_staging_AAAA" {
│
╵
```

epicfaace commented:
@jsf9k Can you try again? I changed the id

jsf9k commented on route53_crossfeed_app.tf:
```diff
@@ -197,36 +197,65 @@ resource "aws_route53_record" "crossfeed_prod_ses_TXT" {
+    zone_id                = "Z33AYJ8TM3BH4J"
+    evaluate_target_health = false
+  }
+  name    = "crossfeed.${aws_route53_zone.cyber_dhs_gov.name}"
```
Should this be something like `staging.crossfeed.${aws_route53_zone.cyber_dhs_gov.name}`?  You already have an `A` record with this name [here](https://github.com/cisagov/cool-dns-cyber.dhs.gov/pull/78/files#diff-5191386bc2ec7103d66596882349fc3029e8e15deee5d010954e4f19ca32fe37R9-R20).

epicfaace replied:
Yeah, thanks for catching that -- updated.

jsf9k replied:
The `A` and `AAAA` records now apply cleanly.

jsf9k commented on route53_crossfeed_app.tf:
```diff
@@ -197,36 +197,65 @@ resource "aws_route53_record" "crossfeed_prod_ses_TXT" {
+    zone_id                = "Z33AYJ8TM3BH4J"
+    evaluate_target_health = false
+  }
+  name    = "crossfeed.${aws_route53_zone.cyber_dhs_gov.name}"
```
Should this be something like `staging.crossfeed.${aws_route53_zone.cyber_dhs_gov.name}`?  You already have an `AAAA` record with this name [here](https://github.com/cisagov/cool-dns-cyber.dhs.gov/pull/78/files#diff-5191386bc2ec7103d66596882349fc3029e8e15deee5d010954e4f19ca32fe37R22-R33).

epicfaace replied:
Yeah, thanks for catching that -- updated.

jsf9k replied:
The `A` and `AAAA` records now apply cleanly.

jsf9k reviewed (commented):
> @jsf9k Can you try again? I changed the id

It looks like the `A` and `AAAA` records failed again, although it takes 45 minutes to finish.  (The `aws_acm_certificate_validation.rules` resource that fails to be created has a _long_ timeout in Terraform.)

I'm not sure, but could the fact that you're creating `A` and `AAAA` records with the same name as the corresponding production records be a problem?

epicfaace commented:
> It looks like the `A` and `AAAA` records failed again, although it takes 45 minutes to finish. (The `aws_acm_certificate_validation.rules` resource that fails to be created has a _long_ timeout in Terraform.)

@jsf9k Thanks for catching. Just updated the record names -- thanks for your help! Note the `aws_acm_certificate_validation.rules` isn't affected by this PR (or related to Crossfeed, I believe).

jsf9k commented:
> Note the `aws_acm_certificate_validation.rules` isn't affected by this PR (or related to Crossfeed, I believe).

You're right.  I'll create an issue for us to look into why that is failing.

dav3r commented:
@epicfaace Can you (or someone else on your team) please ensure that this PR's description aligns with the final changes that were made?

epicfaace commented:
@dav3r I just updated the PR description -- the only additional change that was made was "Added A / AAAA records for staging.crossfeed.cyber.dhs.gov and api.staging.crossfeed.cyber.dhs.gov to point to the LZ.", which I added in a bullet point.

I also just added a small commit (829b87048e160cdfc61d45e0f13c04d834705af8) to use A / AAAA records instead of CNAME records for `api.staging.crossfeed.cyber.dhs.gov`, to be consistent with what we're doing with `staging.crossfeed.cyber.dhs.gov`.

epicfaace commented on route53_crossfeed_app.tf:
```diff
@@ -347,16 +375,6 @@ resource "aws_route53_record" "crossfeed_staging_api_CNAME2" {
-resource "aws_route53_record" "crossfeed_staging_api_CNAME3" {
-  provider = aws.route53resourcechange
-
-  name    = "staging.api.crossfeed.${aws_route53_zone.cyber_dhs_gov.name}"
```
staging.api was also a typo, it's supposed to be api.staging -- just changed in this commit.

dav3r commented on README.md:
```diff
@@ -30,13 +30,13 @@ zone.  This role has a trust relationship with the users account.
 | terraform | n/a |
 
-## Modules ##
+## Modules
```
Unfortunately, `terraform-docs` disagrees with our Markdown linter rule for headings, so this must be changed back to:
```suggestion
## Modules ##
```

dav3r commented on README.md:
```diff
@@ -30,13 +30,13 @@ zone.  This role has a trust relationship with the users account.
 | read\_terraform\_state | github.com/cisagov/terraform-state-read-role-tf-module | n/a |
 
-## Resources ##
+## Resources
```
Unfortunately, `terraform-docs` disagrees with our Markdown linter rule for headings, so this must be changed back to:
```suggestion
## Resources ##
```

dav3r commented on route53_crossfeed_app.tf:
```diff
@@ -271,57 +300,76 @@ resource "aws_route53_record" "crossfeed_staging_TXT" {
+  zone_id = aws_route53_zone.cyber_dhs_gov.zone_id
+}
+
+resource "aws_route53_record" "crossfeed_staging_api_CNAME" {
```
Now that this has all been sorted out, if there's more than one CNAME record, the resource names should be CNAME1, CNAME2, etc:
```suggestion
resource "aws_route53_record" "crossfeed_staging_api_CNAME1" {
```

dav3r commented on README.md:
```diff
@@ -79,16 +79,18 @@ zone.  This role has a trust relationship with the users account.
 | [aws_route53_record.crossfeed_staging_api_A](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_record) | resource |
 | [aws_route53_record.crossfeed_staging_api_AAAA](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_record) | resource |
-| [aws_route53_record.crossfeed_staging_api_CNAME1](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_record) | resource |
+| [aws_route53_record.crossfeed_staging_api_CNAME](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_record) | resource |
```
Now that this has all been sorted out, if there's more than one CNAME record, the resource names should be CNAME1, CNAME2, etc (this change should be batched with the one below):
```suggestion
| [aws_route53_record.crossfeed_staging_api_CNAME1](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_record) | resource |
```

dav3r reviewed (changes requested):
One small resource naming nitpick:

dav3r commented on route53_crossfeed_app.tf:
```diff
@@ -197,27 +197,56 @@ resource "aws_route53_record" "crossfeed_prod_ses_TXT" {
+  zone_id = aws_route53_zone.cyber_dhs_gov.zone_id
+}
+
+
```
Remove unnecessary extra blank line
```suggestion
```

jsf9k commented:
@epicfaace - For future reference, when we accept someone else's suggested changes we try to make the commit message descriptive of the specific changes being made. So for @dav3r's suggestions [here](https://github.com/cisagov/cool-dns-cyber.dhs.gov/pull/78#pullrequestreview-1559279805) we would probably combine the two changes into a single commit with a message something like "Rename a resource".

jsf9k commented:
> > Note the `aws_acm_certificate_validation.rules` isn't affected by this PR (or related to Crossfeed, I believe).
> 
> You're right. I'll create an issue for us to look into why that is failing.

See #79.  (This issue is intended for @cisagov/team-ois not @cisagov/crossfeed.)

dav3r reviewed (changes requested):
I've got a couple of small items.

dav3r commented on route53_crossfeed_app.tf:
```diff
@@ -271,57 +300,76 @@ resource "aws_route53_record" "crossfeed_staging_TXT" {
 
-  name    = "staging.api.crossfeed.${aws_route53_zone.cyber_dhs_gov.name}"
-  records = ["crossfeed-stage-1792947306.us-gov-west-1.elb.amazonaws.com"]
+  name    = "_75f0c44adb3c8ec0023ddc11585227e8.api.staging-cd.crossfeed.${aws_route53_zone.cyber_dhs_gov.name}"
```
Is this meant to be:
`_75f0c44adb3c8ec0023ddc11585227e8.api.staging-cd.crossfeed...` or should it be
`_75f0c44adb3c8ec0023ddc11585227e8.api.staging.crossfeed...` (no "-cd")?

Confirming since the resource name here is `crossfeed_staging_api_CNAME2` (not `crossfeed_staging_cd_api_CNAME2`) and these things match in the case of the other records.

epicfaace replied:
It should have "-cd" in it. I can rename the resource name though to be consistent.

epicfaace replied:
@dav3r Updated the names now.

dav3r replied:
Looks good, thanks!

jsf9k reviewed (approved):
I was able to apply these changes, modulo the issue in #79 that does not concern this PR.  Thanks for your patience, @epicfaace!
